### PR TITLE
ref(hc): Common Option interface/mixin for ORM and RPC models

### PR DESCRIPTION
**Pull request by RyanSkonnord**
Extract from `Organization` and `Project` a mixin for common methods related to options, and have `RpcOrganization` and `RpcProject` implement the same interface.

Introduce `ProjectService`. Add RPC methods for reading, writing, and deleting organization and project options.

### CODE REVIEW

**Review by corps — COMMENTED** [review]
Would it be possible to support multi gets?  In my latest branch, I needed to extend optiosn to support this use case for batch fetching.

**Comment by RyanSkonnord** [comment]
> Would it be possible to support multi gets? In my latest branch, I needed to extend optiosn to support this use case for batch fetching.

Sure, that should work.

File: src/sentry/models/project.py [hunk 1]
```diff
@@ -224,15 +227,19 @@ def is_internal_project(self):
...
-    def get_option(self, *args, **kwargs):
-        return projectoptions.get(self, *args, **kwargs)
+    def update_option(self, key: str, value: Value) -> None:
+        projectoptions.update_rev_for_option(self)
+        return super().update_option(key, value)
 
-    def delete_option(self, *args, **kwargs):
-        return projectoptions.delete(self, *args, **kwargs)
+    def delete_option(self, key: str) -> None:
+        projectoptions.update_rev_for_option(self)
```
Comment (markstory) [inline src/sentry/models/project.py:241]: Longer term should we align and combine the `projectoptions` wrappers with the model APIs? Continuing to have multiple paths to set project options seems like a trap for future us.

File: src/sentry/services/hybrid_cloud/organization/impl.py [hunk 2]
```diff
@@ -464,3 +464,18 @@ def update_region_user(self, *, user: RpcRegionUser, region_name: str) -> None:
         # the attribute we are changing never needs to produce an outbox.
         with in_test_psql_role_override("postgres"):
             OrganizationMember.objects.filter(user_id=user.id).update(user_is_active=user.is_active)
+
+    def get_option(self, *, organization_id: int, key: str) -> OptionValue:
+        orm_organization = Organization.objects.get(id=organization_id)
```
Comment (markstory) [inline src/sentry/services/hybrid_cloud/organization/impl.py:469]:
```suggestion
        orm_organization = Organization.objects.get_from_cache(id=organization_id)
```
Option reads don't use much organization state so we should be able to read those records from cache.

File: src/sentry/services/hybrid_cloud/organization/impl.py [hunk 3]
```diff
@@ -464,3 +464,18 @@ def update_region_user(self, *, user: RpcRegionUser, region_name: str) -> None:
...
+        value = orm_organization.get_option(key)
+        if value is not None and not isinstance(value, (str, int, bool)):
+            raise TypeError
+        return value
+
+    def update_option(self, *, organization_id: int, key: str, value: OptionValue) -> bool:
+        orm_organization = Organization.objects.get(id=organization_id)
```
Comment (markstory) [inline src/sentry/services/hybrid_cloud/organization/impl.py:476]:
```suggestion
        orm_organization = Organization.objects.get_from_cache(id=organization_id)
```

File: src/sentry/services/hybrid_cloud/organization/impl.py [hunk 4]
```diff
@@ -464,3 +464,18 @@ def update_region_user(self, *, user: RpcRegionUser, region_name: str) -> None:
...
+        return orm_organization.update_option(key, value)  # type: ignore[no-any-return]
+
+    def delete_option(self, *, organization_id: int, key: str) -> None:
+        orm_organization = Organization.objects.get(id=organization_id)
```
Comment (markstory) [inline src/sentry/services/hybrid_cloud/organization/impl.py:480]:
```suggestion
        orm_organization = Organization.objects.get_from_cache(id=organization_id)
```

File: src/sentry/services/hybrid_cloud/project/impl.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+from sentry.models import Project, ProjectOption
+from sentry.services.hybrid_cloud import OptionValue
+from sentry.services.hybrid_cloud.project import ProjectService, RpcProject, RpcProjectOptionValue
+
+
+class DatabaseBackedProjectService(ProjectService):
+    def get_option(self, *, project: RpcProject, key: str) -> RpcProjectOptionValue:
+        from sentry import projectoptions
+
+        orm_project = Project.objects.get(id=project.id)
```
Comment (markstory) [inline src/sentry/services/hybrid_cloud/project/impl.py:12]:
```suggestion
        orm_project = Project.objects.get_from_cache(id=project.id)
```

We should be able to read from cache here as reading/writing options doesn't use much project state.

File: src/sentry/models/project.py (hunk 1 above)
Comment (RyanSkonnord) [reply src/sentry/models/project.py:241]: I agree.

**Review by markstory — APPROVED** [review]
Looks good to me.

**Pull request merged**